### DISCUSSION TITLE
Include a sub item based on a shared field in the main table

### DISCUSSION THREAD

**Discussion by noelzappy** [prisma/prisma#25792] [Q&A]
### Question

```
model Item {
  id           Int           @id @default(autoincrement())
  name         String
  category     String
  stateId      Int
  state        ItemState     @relation(fields: [stateId], references: [id])
  relatedItems RelatedItem[] @relation("ItemRelatedItems")
}

model RelatedItem {
  id       Int       @id @default(autoincrement())
  name     String
  category String
  itemId   Int
  stateId  Int
  state    ItemState @relation(fields: [stateId], references: [id])
  item     Item      @relation("ItemRelatedItems", fields: [itemId], references: [id])
}

model ItemState {
  id          Int           @id @default(autoincrement())
  name        String
  Item        Item[]
  RelatedItem RelatedItem[]
}
```

These are three Prisma models. I want to run a `findMany` query on `Items` and include only records from `RelatedItem` that have the same `stateId` as the `Item`. The `stateId` will not be passed as a variable to the query. I could not find anywhere in the docs that describe this. How can I do this?

### Information about Prisma Schema, Client Queries and Environment (optional)

```
model Item {
  id           Int           @id @default(autoincrement())
  name         String
  category     String
  stateId      Int
  state        ItemState     @relation(fields: [stateId], references: [id])
  relatedItems RelatedItem[] @relation("ItemRelatedItems")
}

model RelatedItem {
  id       Int       @id @default(autoincrement())
  name     String
  category String
  itemId   Int
  stateId  Int
  state    ItemState @relation(fields: [stateId], references: [id])
  item     Item      @relation("ItemRelatedItems", fields: [itemId], references: [id])
}

model ItemState {
  id          Int           @id @default(autoincrement())
  name        String
  Item        Item[]
  RelatedItem RelatedItem[]
}
```

**Comment by ludralph** [comment]
Hi @noelzappy 

The standard `include` option doesn't allow for filtering related records based on a comparison with the parent record's fields. While you can filter included relations, the filtering is typically based on fixed values or parameters, not on values from the parent record. To achieve your usecase,  you might need to use a raw SQL query.

**Comment by steventete** [comment]
To achieve the desired behavior—filtering `RelatedItem` records to include only those with the same `stateId` as the `Item` in a `findMany` query—you can use Prisma's `select` and `include` options with nested filtering logic.

### Query Example

```javascript
const items = await prisma.item.findMany({
  include: {
    relatedItems: {
      where: {
        stateId: {
          equals: prisma.item.stateId, // Match `stateId` of `RelatedItem` to `Item`
        },
      },
    },
  },
});
```
## Explanation

1. `findMany` on Item: Fetches all Item records.
2. `include`: Includes the relatedItems relation in the results.
3. `where` filter inside relatedItems: Ensures only RelatedItem records with a stateId equal to the stateId of the parent Item are included.

However, direct filtering like this (referencing item.stateId dynamically) is not currently supported by Prisma out-of-the-box. Prisma does not allow accessing parent model properties in nested filters directly.
Alternative Solution: Post-Processing in JavaScript

Since Prisma doesn't support the dynamic parent-child filtering within the query, you can handle it by retrieving all related records and filtering them in your application code.

```javascript
const items = await prisma.item.findMany({
  include: {
    relatedItems: true,
  },
});

const filteredItems = items.map(item => ({
  ...item,
  relatedItems: item.relatedItems.filter(relatedItem => relatedItem.stateId === item.stateId),
}));
```
### Key Points

- Limitations: Currently, Prisma does not natively allow referencing parent fields in nested where filters.
- Workaround: Use application-level filtering after fetching the data.
